### PR TITLE
Implement CXF and Backup and Restore Logic

### PR DESCRIPTION
**Pull request by AdityaMitra5102**
Implemented CXF and Backup and Restore Logic

Changelog:

- Added nitrokey/nk3/credential_exchange_format.py with partial implementation of CXF. Did not implement types not required like address, credit card, etc. Followed https://fidoalliance.org/specs/cx/cxf-v1.0-ps-errata-20260309.html
- Took out the Dataclasses from secrets_app.py and added them to secrets_app_dataclasses.py to prevent circular import
- Added backup and restore logic to secrets_app.py

### CODE REVIEW

File: src/nitrokey/nk3/credential_exchange_format.py [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+            ],
+            timestamp=uint(time.time()),
+        )
+
+    @classmethod
+    def cxf_to_items(cls, payload: CXFPayload) -> List[Item]:
+        return (
+            payload.accounts[0].items
+            if payload and payload.accounts and payload.accounts[0]
+            else []
```
Comment (sosthene-nitrokey) [inline src/nitrokey/nk3/credential_exchange_format.py]: It might make sense to merge the multiple accounts together if there are many don't you think?

So that the app can import third-party exports of accounts.

File: src/nitrokey/nk3/credential_exchange_format.py [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+                    urls=[],  # Nitrokey does not store URLs
+                    username=EditableField(
+                        id=_get_random_id(),
+                        fieldType=FieldType.STRING,
+                        value=pse.login.decode("utf-8", errors="ignore") if pse.login else "",
+                    ),
+                    password=EditableField(
+                        id=_get_random_id(),
+                        fieldType=FieldType.CONCEALED_STRING,
+                        value=pse.password.decode("utf-8", errors="ignore") if pse.password else "",
```
Comment (sosthene-nitrokey) [inline src/nitrokey/nk3/credential_exchange_format.py]: If there is no password it likely means that it's a TOTP/HOTP credential and I think they should not be encoded as `BasicAuth` and there should be a warning in the UX that such credentials cannot be exported so that users don't lose data without realizing it until too late.

File: src/nitrokey/nk3/credential_exchange_format.py (hunk 1 above)
Comment (AdityaMitra5102) [reply src/nitrokey/nk3/credential_exchange_format.py]: Yes, I'll make the required changes on this

File: src/nitrokey/nk3/credential_exchange_format.py (hunk 2 above)
Comment (AdityaMitra5102) [reply src/nitrokey/nk3/credential_exchange_format.py]: Yes. I assumed TOTP and HOTP cannot be exported and hence, only passwords are being exported. So I encoded everything as basic auth. The 'if password is none' clause is mainly for the Static Typing checkers, and that part would realistically never be invoked. 

About warning on the UX, I guess I should be adding that into the Nitrokey App2 and Nitropy and not the SDK. Is my reasoning correct?

Comment (sosthene-nitrokey) [reply src/nitrokey/nk3/credential_exchange_format.py]: Yes, but I think the way it should be done is that the warning is not shown systematically, but instead shown only when there are credentials that cannot be exported (and the UX might even show which credentials).

So for that the SDK should also return the information of which/how many credentials could not be exported.

Comment (AdityaMitra5102) [reply src/nitrokey/nk3/credential_exchange_format.py]: Okay. I shall take that into consideration and try to implement it

Comment (AdityaMitra5102) [reply src/nitrokey/nk3/credential_exchange_format.py]: I am returning a list of labels of the items which could not be exported. Then we can handle it in the app and nitropy

File: src/nitrokey/nk3/secrets_app_dataclasses.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
+# Moved the dataclasses from secrets_app to here to prevent circular import when trying to import in CXF
```
Comment (robin-nitrokey) [inline src/nitrokey/nk3/secrets_app_dataclasses.py]: I would prefer to keep the dataclasses in `secrets_app.py` for a clearer code structure. To avoid the circular imports, you can move the imports into the relevant functions and use strings and conditional imports for the type annotations like we do here:
- https://github.com/Nitrokey/nethsm-sdk-py/blob/081048b3c55e9eff0a2fa5915df88cffdaa99e95/nethsm/__init__.py#L36
- https://github.com/Nitrokey/nethsm-sdk-py/blob/081048b3c55e9eff0a2fa5915df88cffdaa99e95/nethsm/__init__.py#L564

File: src/nitrokey/nk3/credential_exchange_format.py [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+
+    @classmethod
+    def cxf_from_dict(cls, d: dict[str, Any]) -> CXFPayload:
+        accounts = []
+        for acc in d.get("accounts", []):
+            items = []
+            for item_d in acc.get("items", []):
+                credentials = []
+                for cred in item_d.get("credentials", []):
+                    credentials.append(cls.basic_auth_from_dict(cred))
```
Comment (robin-nitrokey) [inline src/nitrokey/nk3/credential_exchange_format.py]: Don’t we have to check that this really is a Basic Auth credential before parsing it?

File: src/nitrokey/nk3/credential_exchange_format.py [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+                extensions = []
+                for ext in item_d.get("extensions", []):
+                    extensions.append(cls.nitrokey_password_extension_from_dict(ext))
```
Comment (robin-nitrokey) [inline src/nitrokey/nk3/credential_exchange_format.py]: Again I think we have to check that this is really the Nitrokey extension.

**Review by robin-nitrokey — COMMENTED** [review]
Thanks for the PR! I’ve added some initial comments. Can you also please add some test cases for the conversion in both directions?

File: src/nitrokey/nk3/secrets_app_dataclasses.py (hunk 3 above)
Comment (AdityaMitra5102) [reply src/nitrokey/nk3/secrets_app_dataclasses.py]: Fixed this

File: src/nitrokey/nk3/credential_exchange_format.py (hunk 4 above)
Comment (AdityaMitra5102) [reply src/nitrokey/nk3/credential_exchange_format.py]: I checked the existence of a password field to ensure it is a Basic Auth

File: src/nitrokey/nk3/credential_exchange_format.py (hunk 5 above)
Comment (AdityaMitra5102) [reply src/nitrokey/nk3/credential_exchange_format.py]: I checked the 'name' field of the extension to confirm whether its the Nitrokey extension

**Comment by AdityaMitra5102** [comment]
Added testing the export encryption in the unittest as well

File: src/nitrokey/nk3/credential_exchange_format.py [hunk 6]
```diff
@@ -0,0 +1,411 @@
...
+
+@dataclass
+class Collection:
+    id: b64url
+    title: tstr
+    items: List[LinkedItem]
+    creationAt: Optional[uint] = None
+    modifiedAt: Optional[uint] = None
+    subtitle: Optional[tstr] = None
+    icon: Optional[tstr] = None
```
Comment (sosthene-nitrokey) [inline src/nitrokey/nk3/credential_exchange_format.py]: Not in the spec
```suggestion
```

File: src/nitrokey/nk3/credential_exchange_format.py [hunk 7]
```diff
@@ -0,0 +1,411 @@
...
+    API_KEY = "api-key"
+    BASIC_AUTH = "basic-auth"
+    CREDIT_CARD = "credit-card"
+    CUSTOM_FIELDS = "custom-fields"
+    DRIVERS_LICENSE = "drivers-license"
+    FILE = "file"
+    GENERATED_PASSWORD = "generated-password"
+    PASSKEY = "passkey"
+    TOTP = "totp"
+    # TODO Add the rest of them
```
Comment (sosthene-nitrokey) [inline src/nitrokey/nk3/credential_exchange_format.py]: What is the decision process for including or excluding them?

We don't have specific support for address, credit-card, drivers license, files so I think either we should have all of them or only the ones that we support.

I'm in favour of including all the ones from the spec.

File: src/nitrokey/nk3/secrets_app.py [hunk 8]
```diff
@@ -278,12 +311,6 @@ def from_attribute_byte_type(cls, a: int) -> "Kind":
 }
 
 
-class Algorithm(Enum):
-    Sha1 = 0x01
-    Sha256 = 0x02
-    Sha512 = 0x03
-
-
```
Comment (sosthene-nitrokey) [inline src/nitrokey/nk3/secrets_app.py:286]: Is there a specific reason why you re-ordered this ? Please keep them in the same order to keep the diff small and `git blame` good.

File: src/nitrokey/nk3/secrets_app.py [hunk 9]
```diff
@@ -479,6 +506,82 @@ def get_credential(self, cred_id: bytes) -> PasswordSafeEntry:
...
+                if item.kind != Kind.NotSet:
+                    non_exportable_credentials.append(item.label)
+                else:
+                    export_list.append(
+                        PasswordToCXF.password_to_item(item, self.get_credential(item.label))
+                    )
+            except SecretsAppException as e:
+                self.logfn(f"Exception in getting credential -> {e}")
+                non_exportable_credentials.append(item.label)
+                continue  # Incorrect pin will still retrieve credentials not protected by pin
```
Comment (sosthene-nitrokey) [inline src/nitrokey/nk3/secrets_app.py]: I think if we want to support exporting without the PIN, we should do it by making the password optional, and skipping PIN protected passwords.

If the PIN is present and wrong, it should error not skip the PIN protected ones

File: src/nitrokey/nk3/secrets_app.py [hunk 10]
```diff
@@ -479,6 +506,82 @@ def get_credential(self, cred_id: bytes) -> PasswordSafeEntry:
...
+        return export_list, non_exportable_credentials
+
+    def get_export_cxf(
+        self, password: str = "", as_dict: bool = False
```
Comment (sosthene-nitrokey) [inline src/nitrokey/nk3/secrets_app.py]: What is the purpose of this `as_dict` ? 

I don't like such dependent types.

File: src/nitrokey/nk3/credential_exchange_format.py [hunk 11]
```diff
@@ -0,0 +1,411 @@
...
+        groups = [chars[i : i + 5] for i in range(0, 25, 5)]
+        return "-".join("".join(g) for g in groups)
+
+    @classmethod
+    def use_passphrase(cls, passphrase: str) -> EncryptCXF:
+        passphrase = passphrase.strip().replace("-", "").replace(" ", "").upper()
+        hkdf = HKDF(
+            algorithm=hashes.SHA256(),
+            length=32,
+            salt=b"\x99\xf2L\x93\xe1\xe3%6\x0f}\x17\x0f",  # Random but constant salt
```
Comment (sosthene-nitrokey) [inline src/nitrokey/nk3/credential_exchange_format.py]:
```suggestion
            salt=b"\xc8*\xdf\xf6Ac\xb9|S'u\xb0\x98\x19b@",  # Random but constant salt
```

Your salt is 12 bytes, this one is 16 bytes, so we have the expected uniform randomness of 128 bits.

File: src/nitrokey/nk3/credential_exchange_format.py [hunk 12]
```diff
@@ -0,0 +1,411 @@
...
+            info=None,
```
Comment (sosthene-nitrokey) [inline src/nitrokey/nk3/credential_exchange_format.py]:
```suggestion
            info=b"nitrokey-export-encryption-key",
```

**Review by sosthene-nitrokey — CHANGES_REQUESTED** [review]
Some small nits, besides that LGTM.

Please also add a test that deserializes an example CXF payload (such as the one in [Appendix A](https://fidoalliance.org/specs/cx/cxf-v1.0-ps-errata-20260309.html#appendix-a))

**Comment by robin-nitrokey** [comment]
I’ve added some suggested changes regarding the code structure here: https://github.com/AdityaMitra5102/nitrokey-sdk-py/pull/1. These also address some of the issues Sosthène raised. Please take a look.

**Comment by robin-nitrokey** [comment]
Regarding the functionality:
- I think we need to make the parsing a bit more robust, i. e. add more validation, for example to ensure that fields have the correct type.
- As Sosthène mentioned, we should test against the example from the spec and also add some more fixed examples, e. g. with unsupported extensions, missing username or password and other edge cases.
- You mention that you chose the encrypted format for downstream compatibility. What is the downstream in this context?

File: src/nitrokey/nk3/credential_exchange_format.py (hunk 7 above)
Comment (AdityaMitra5102) [reply src/nitrokey/nk3/credential_exchange_format.py]: I'll add all

File: src/nitrokey/nk3/credential_exchange_format.py (hunk 6 above)
Comment (AdityaMitra5102) [reply src/nitrokey/nk3/credential_exchange_format.py]: Will remove it. I wrote that because I was confused among different versions of the spec opened accidentally

File: src/nitrokey/nk3/secrets_app.py (hunk 8 above)
Comment (AdityaMitra5102) [reply src/nitrokey/nk3/secrets_app.py:286]: I had moved them out to a different file to remove circular dependency, followed by adding them back as per Robin's comment at https://github.com/Nitrokey/nitrokey-sdk-py/pull/126#discussion_r3364514923. I'll fix the order back.

File: src/nitrokey/nk3/credential_exchange_format.py (hunk 4 above)
Comment (AdityaMitra5102) [reply src/nitrokey/nk3/credential_exchange_format.py]: Changed it to check credential type directly

**Comment by AdityaMitra5102** [comment]
I mentioned returning as **dict** with B64 Encoded ciphertext for downstream compatibility. Which ensures we can use the same json.dumps and json.loads functions for both encrypted and cleartext backups without having to define separate procedures to store and read from file

**Review by sosthene-nitrokey — APPROVED** [review]

File: src/nitrokey/nk3/credential_exchange_format.py [hunk 13]
```diff
@@ -0,0 +1,404 @@
...
+            exporterDisplayName=d.get("exporterDisplayName", ""),
+            timestamp=d.get("timestamp", 0),
+            accounts=accounts,
+        )
+
+    @staticmethod
+    def from_items(items: List[Item]) -> Header:
+        return Header(
+            version=Version(1, 0),  # v1.0
+            exporterRpId="nitrokey",  # Exporting from authenticator
```
Comment (robin-nitrokey) [inline src/nitrokey/nk3/credential_exchange_format.py]: RP IDs are typically existing domain names so I think it would be better to use `nitrokey.com`.

File: src/nitrokey/nk3/credential_exchange_format.py [hunk 14]
```diff
@@ -0,0 +1,404 @@
...
+
+    def encrypt(self, key: CXFKey) -> dict[str, Any]:
+        payload_bytes = json.dumps(asdict(self)).encode()
+        aesgcm = AESGCM(key._key)
+        nonce = secrets.token_bytes(12)
+        ct = aesgcm.encrypt(nonce, payload_bytes, None)
+        encrypted = ct + nonce
+        return {
+            "EncryptedCXF": urlsafe_b64encode(encrypted).decode()
+        }  # Returning as dict to maintain downstream compatibility
```
Comment (robin-nitrokey) [inline src/nitrokey/nk3/credential_exchange_format.py:99]: What do you think about including a version field in case we want to change the format or encryption scheme later?

File: src/nitrokey/nk3/credential_exchange_format.py [hunk 15]
```diff
@@ -0,0 +1,404 @@
...
+
+    @staticmethod
+    def from_dict(item_d: dict[str, Any]) -> Item:
+        credentials = []
+        for cred in item_d.get("credentials", []):
+            if cred["type"] == CredentialType.BASIC_AUTH:  # Check whether cred is BasicAuth
+                credentials.append(BasicAuth.from_dict(cred))
+        extensions = []
+        for ext in item_d.get("extensions", []):
+            if ext["name"] == "Nitrokey Password Extension":
```
Comment (robin-nitrokey) [inline src/nitrokey/nk3/credential_exchange_format.py]: Please use a constant here and in `NitrokeyPasswordExtension`.

According to the spec, extension names should follow a special scheme:

> If this is a custom extension defined by the exporter, then the value MUST take the following format: EXPORTER_RP_ID/EXTENSION_NAME.

So maybe use `nitrokey.com/SecretsAppMetadata`?

File: src/nitrokey/nk3/credential_exchange_format.py [hunk 16]
```diff
@@ -0,0 +1,404 @@
...
+            label=d.get("label"),
+            designation=d.get("designation"),
+        )
+
+
+@dataclass
+class BasicAuth(Credential):
+    urls: List[tstr]
+    username: EditableField
+    password: EditableField
```
Comment (robin-nitrokey) [inline src/nitrokey/nk3/credential_exchange_format.py]: Both fields should be optional according to the spec.

File: src/nitrokey/nk3/secrets_app.py [hunk 17]
```diff
@@ -479,6 +515,83 @@ def get_credential(self, cred_id: bytes) -> PasswordSafeEntry:
         p.properties = p.properties.hex().encode() if p.properties else None
         return p
 
+    def export_cxf(self, encryption: bool, password: str = "") -> CXFBackupCombined:
```
Comment (robin-nitrokey) [inline src/nitrokey/nk3/secrets_app.py]: Again I’m in favor of just returning a `CFXPayload` and optionally do the encryption in a second step (in the calling application). This makes the API simpler and also gives the user more flexibility (e. g. use a given passphrase instead of generating it). What’s the benefit of combining it here?

File: src/nitrokey/nk3/secrets_app.py [hunk 18]
```diff
@@ -479,6 +515,83 @@ def get_credential(self, cred_id: bytes) -> PasswordSafeEntry:
...
+            )
+        else:
+            passphrase = CXFKey.generate_passphrase()
+            return CXFBackupCombined(
+                payload=cxfpayload.encrypt(CXFKey.use_passphrase(passphrase)),
+                skipped_credentials=non_exportable_credentials,
+                passphrase=passphrase,
+            )
+
+    def import_cxf(self, import_cxf_combined: CXFBackupCombined, password: str = "") -> None:
```
Comment (robin-nitrokey) [inline src/nitrokey/nk3/secrets_app.py]: This should also return a result dataclass with the names of the imported credentials and also the failed credentials so that applications can show a summary of the import to the user.

File: src/nitrokey/nk3/secrets_app.py (hunk 17 above)
Comment (robin-nitrokey) [inline src/nitrokey/nk3/secrets_app.py]: For both this and the import function: What do you think about adding an optional progress callback that is called in the beginning and after every import/export with the total number of credentials and the number of processed credentials? Then applications could show a progress bar. I think that could be useful especially for larger imports and exports.

**Review by robin-nitrokey — COMMENTED** [review]
Thanks for the update! I’ve added some comments.

File: src/nitrokey/nk3/credential_exchange_format.py (hunk 13 above)
Comment (AdityaMitra5102) [reply src/nitrokey/nk3/credential_exchange_format.py]: Fixed

File: src/nitrokey/nk3/credential_exchange_format.py (hunk 14 above)
Comment (AdityaMitra5102) [reply src/nitrokey/nk3/credential_exchange_format.py:99]: Should I make it a static number? Or do I take it from importlib.metadata.version("nitrokey")???

File: src/nitrokey/nk3/credential_exchange_format.py (hunk 15 above)
Comment (AdityaMitra5102) [reply src/nitrokey/nk3/credential_exchange_format.py]: fixed

File: src/nitrokey/nk3/credential_exchange_format.py (hunk 14 above)
Comment (robin-nitrokey) [reply src/nitrokey/nk3/credential_exchange_format.py:99]: Yes, just `"version": 1,` for the time being. We can then bump it if we ever change the format. And the decrypt function should make sure that `version` is 1 and raise an error otherwise.

File: src/nitrokey/nk3/credential_exchange_format.py (hunk 16 above)
Comment (AdityaMitra5102) [reply src/nitrokey/nk3/credential_exchange_format.py]: Fixed

File: src/nitrokey/nk3/secrets_app.py (hunk 17 above)
Comment (AdityaMitra5102) [reply src/nitrokey/nk3/secrets_app.py]: About using user defined passphrase, as per a previous discussion with Sosthene and Jan Knittel on Zulip, we decided generating one gives us 128 bits of security over user-provided passphrases. Hence, I embedded them into a single function. If needed, I would modify this function to break the process up.

File: src/nitrokey/nk3/secrets_app.py (hunk 18 above)
Comment (AdityaMitra5102) [reply src/nitrokey/nk3/secrets_app.py]: fixed

File: src/nitrokey/nk3/secrets_app.py (hunk 17 above)
Comment (AdityaMitra5102) [reply src/nitrokey/nk3/secrets_app.py]: During import, the total credentials might be present in a nested order and hence we are not getting total credentials in the beginning. To implement this, we would need to loop through it twice to count the total credentials first. 

Currently it is performing it like:
```
for cxfitem in item_list:
            pr_list = cxfitem.password_representation()
            for pr in pr_list:
        
```

File: src/nitrokey/nk3/credential_exchange_format.py (hunk 14 above)
Comment (AdityaMitra5102) [reply src/nitrokey/nk3/credential_exchange_format.py:99]: fixed

File: src/nitrokey/nk3/credential_exchange_format.py (hunk 15 above)
Comment (AdityaMitra5102) [reply src/nitrokey/nk3/credential_exchange_format.py]: If it is 
> EXPORTER_RP_ID/EXTENSION_NAME

Shouldn't we use `nitrokey.com/NitrokeyPasswordExtension` instead of `nitrokey.com/SecretsAppMetadata` since that is the extension name?

Comment (robin-nitrokey) [reply src/nitrokey/nk3/credential_exchange_format.py]: Well, it is up to us to define the extension name. :) I proposed `SecretsAppMetadata` because I find it more descriptive of the actual content and it fits the example name from the spec (`VaultType`) more closely. But I don’t have a strong opinion on it.

Comment (AdityaMitra5102) [reply src/nitrokey/nk3/credential_exchange_format.py]: Okay, I'll rename the extension

Comment (AdityaMitra5102) [reply src/nitrokey/nk3/credential_exchange_format.py]: Fixed

File: src/nitrokey/nk3/secrets_app.py (hunk 17 above)
Comment (robin-nitrokey) [reply src/nitrokey/nk3/secrets_app.py]: As `item_list` is already parsed and in memory, it should be no problem to count all credentials before starting the import. Alternatively, showing the progress in terms of items instead of credentials would also be better than nothing.

Comment (AdityaMitra5102) [reply src/nitrokey/nk3/secrets_app.py]: Okay, I'll add that.

Comment (AdityaMitra5102) [reply src/nitrokey/nk3/secrets_app.py]: Added. I am using this definition `Callable[[int, CXFRestoreCombined], None]`. Hence instead of keeping a count in the SDK, the caller may run `len()` function on the arrays in the dataclass. It would also allow the caller to show a live progress with exact names if required in future. The first `int` contains the total number of credentials.

Comment (robin-nitrokey) [reply src/nitrokey/nk3/secrets_app.py]: Looks good, thank you! What about the import?

Comment (AdityaMitra5102) [reply src/nitrokey/nk3/secrets_app.py]: This is in the import function. Do I also add a callback for export?

Comment (robin-nitrokey) [reply src/nitrokey/nk3/secrets_app.py]: Sorry, yes, I meant the export.

Comment (AdityaMitra5102) [reply src/nitrokey/nk3/secrets_app.py]: Ok, I'll add one in the export

Comment (AdityaMitra5102) [reply src/nitrokey/nk3/secrets_app.py]: Added callback to export

Comment (robin-nitrokey) [reply src/nitrokey/nk3/secrets_app.py]: Thanks!

File: src/nitrokey/nk3/secrets_app.py [hunk 19]
```diff
@@ -479,6 +522,119 @@ def get_credential(self, cred_id: bytes) -> PasswordSafeEntry:
...
+                if item.properties.secret_encryption and password:
+                    self.verify_pin_raw(password)
+
+                try:
+                    self.register(
+                        credid=item.label,
+                        kind=item.kind,
+                        algo=item.algorithm,
+                        touch_button_required=item.properties.touch_required,
+                        pin_based_encryption=item.properties.secret_encryption,
```
Comment (robin-nitrokey) [inline src/nitrokey/nk3/secrets_app.py:617]: What happens if you try to import an item that has this flag but you don’t provide a password (or the password is not even set)?

File: src/nitrokey/nk3/secrets_app.py (hunk 17 above)
Comment (robin-nitrokey) [reply src/nitrokey/nk3/secrets_app.py]: Separating it would just make us more flexible without adding much complexity (if at all) while also making the type annotations simpler. But it’s not a big issue and we can also go forward with this implementation.

**Review by robin-nitrokey — APPROVED** [review]
Thanks for the replies and fixes! Please squash the commits, then we can merge git.

File: src/nitrokey/nk3/secrets_app.py (hunk 19 above)
Comment (AdityaMitra5102) [reply src/nitrokey/nk3/secrets_app.py:617]: It goes into the failed list

**Comment by AdityaMitra5102** [comment]
I squashed all commits to one

File: src/nitrokey/nk3/secrets_app.py [hunk 20]
```diff
@@ -941,6 +933,7 @@ def change_pin_raw(self, password: str, new_password: str) -> None:
     def verify_pin_raw(self, password: str) -> None:
         structure = [tlv8.Entry(Tag.Password.value, password)]
         self._send_receive(Instruction.VerifyPIN, structure=structure)
+        self.dev.secrets_pin_cache = int(time.time() * 1000)
```
Comment (robin-nitrokey) [inline src/nitrokey/nk3/secrets_app.py]: Please use `datetime.now()` for easier readability.

File: src/nitrokey/trussed/_device.py [hunk 21]
```diff
@@ -148,14 +149,13 @@ def _call_admin_ccid_legacy(
...
-        data: bytes = b"",
-        perform_select: Optional[bool] = True,
-    ) -> bytes:
+    def _call_ccid(self, app: App, response_len: Optional[int] = None, data: bytes = b"") -> bytes:
         assert not isinstance(self.device, CtapHidDevice)
+        perform_select = (
+            (int(time.time() * 1000) - self.secrets_pin_cache) > 100
+            if hasattr(self, "secrets_pin_cache")
+            else True
+        )  # Check 100ms
```
Comment (robin-nitrokey) [inline src/nitrokey/trussed/_device.py]: - I think `(datetime.now() - self.secrets_pin_cache) > timedelta(milliseconds=100)` would be clearer.
- Please initialize the field with `None` in `__init__` (and reset it to `None` instead of -1) and check that instead of `hasattr`.
- I think this should also check whether `app == App.SECRETS`. Otherwise commands could be sent to the wrong application because of the missing select.

File: src/nitrokey/nk3/secrets_app.py (hunk 20 above)
Comment (AdityaMitra5102) [reply src/nitrokey/nk3/secrets_app.py]: fixed

File: src/nitrokey/trussed/_device.py (hunk 21 above)
Comment (AdityaMitra5102) [reply src/nitrokey/trussed/_device.py]: Fixed the first and third.

I tried with setting it in init. It is not being set. Possibly this is due to inheritance and SecretsApp calls takes a NK3 object as device and, which even if it inherits TrussedDevice, may not run the constructor for some reason. 

I have replaced -1 to None accordingly

**Comment by robin-nitrokey** [comment]
@AdityaMitra5102 To make this ready to merge, please rebase onto main, remove the reference to `is_ipc_available` (this should go to a separate PR) and squash the commits where it makes sense. And please add the change to the changelog and the docs.

**Comment by AdityaMitra5102** [comment]
@robin-nitrokey could you take a look now

**Comment by robin-nitrokey** [comment]
@AdityaMitra5102 I think the changelog and docs updates are still missing, no? Otherwise it looks good.

**Comment by AdityaMitra5102** [comment]
I'll add them. I forgot about that.

**Comment by robin-nitrokey** [comment]
Looks good, thank you! I’ll merge once the CI has passed.
